File: src/runners/deepseek.ts
/**
 * DeepSeekRunner — bring-your-own-key (BYOK) DeepSeek cloud backend (§16).
 *
 * Talks to DeepSeek's OpenAI-compatible /chat/completions endpoint with the
 * same in-process MCP servers a real model uses, so ACLs, approval gates and
 * the inbox mechanic apply identically. Agents whose model name is a
 * deepseek-* model are routed here when DEEPSEEK_API_KEY is present.
 *
 * Tool names are sanitized to OpenAI function-name rules ([a-zA-Z0-9_-]+),
 * since MCP names like "tasks.set_status" contain dots that OpenAI rejects.
 */
import type { Runner, RunnerHandle, SessionOutcome } from "./types.js";
import type { McpRuntime, McpServer, McpTool } from "../mcp/context.js";
import type { SessionManifest } from "../domain/types.js";
import type { AgentScript } from "./script.js";
import { serversForManifest } from "../mcp/index.js";
import { jsonSchemaOf } from "../mcp/context.js";
import { buildAgentPrompt } from "./prompt.js";
import { estimateDeepseekCost } from "./models.js";

const MAX_TURNS = 50;

interface DeepseekToolCall {
  id: string;
  type: "function";
  function: { name: string; arguments: string };
}

interface DeepseekMessage {
  role: "system" | "user" | "assistant" | "tool";
  content: string | null;
  tool_calls?: DeepseekToolCall[];
  tool_call_id?: string;
}

export interface DeepseekRunnerOptions {
  apiKey: string;
  baseUrl?: string;
}

/** MCP tool name → OpenAI-safe function name. */
export function sanitizeToolName(name: string): string {
  return name.replace(/[^a-zA-Z0-9_-]/g, "_");
}

export class DeepseekRunner implements Runner {
  kind = "deepseek";

  constructor(private opts: DeepseekRunnerOptions) {}

  async provision(opts: {
    sessionId: string;
    manifest: SessionManifest;
    runtime: McpRuntime;
    script?: AgentScript;
    cwd: string;
  }): Promise<RunnerHandle> {
    const { manifest, runtime } = opts;
    const servers = serversForManifest(manifest.mcpConnections);
    const baseUrl = (this.opts.baseUrl ?? "https://api.deepseek.com").replace(/\/$/, "");
    const model = manifest.agent.model;

    // Build OpenAI function specs + a reverse map back to (server, tool).
    const toolByFn = new Map<string, { server: McpServer; tool: McpTool }>();
    const tools: unknown[] = [];
    for (const server of servers) {
      for (const tool of server.tools) {
        const fn = sanitizeToolName(tool.name);
        toolByFn.set(fn, { server, tool });
        tools.push({
          type: "function",
          function: {
            name: fn,
            description: tool.description,
            parameters: jsonSchemaOf(tool.inputSchema),
          },
        });
      }
    }

    const messages: DeepseekMessage[] = [
      { role: "system", content: buildAgentPrompt(manifest) },
      { role: "user", content: "Begin your work now. Use the available tools, verify your results, and finish when your exit criteria are met." },
    ];

    let resolveDone!: (o: SessionOutcome) => void;
    const done = new Promise<SessionOutcome>((r) => (resolveDone = r));
    let aborted = false;
    let usage = { promptTokens: 0, completionTokens: 0 };

    const fail = (error: string) =>
      resolveDone({ status: "failed", summary: null, costUsd: null, commitShas: [], error });

    const runLoop = async (): Promise<void> => {
      for (let turn = 0; turn < MAX_TURNS && !aborted; turn++) {
        const resp = await this.chat(baseUrl, model, messages, tools);
        const choice = resp.choices?.[0];
        if (!choice) throw new Error("DeepSeek returned no completion choice");
        const msg = choice.message ?? {};
        usage.promptTokens += resp.usage?.prompt_tokens ?? 0;
        usage.completionTokens += resp.usage?.completion_tokens ?? 0;

        const content: string = msg.content ?? "";
        const toolCalls: DeepseekToolCall[] = msg.tool_calls ?? [];

        if (toolCalls.length === 0) {
          resolveDone({
            status: "ok",
            summary: content || "finished",
            costUsd: estimateDeepseekCost(model, usage),
            commitShas: [],
            error: null,
          });
          return;
        }

        messages.push({ role: "assistant", content, tool_calls: toolCalls });

        for (const call of toolCalls) {
          if (aborted) return;
          const entry = toolByFn.get(call.function.name);
          let args: Record<string, unknown> = {};
          try {
            args = call.function.arguments ? (JSON.parse(call.function.arguments) as Record<string, unknown>) : {};
          } catch {
            args = {};
          }

          let output: unknown = null;
          let error: string | null = null;
          if (!entry) {
            error = `unknown tool ${call.function.name}`;
          } else {
            try {
              output = await entry.server.call(runtime, entry.tool.name, args);
            } catch (e) {
              error = e instanceof Error ? e.message : String(e);
            }
          }
          await runtime.recordToolCall({
            ts: new Date().toISOString(),
            name: entry?.tool.name ?? call.function.name,
            input: args,
            output: error ? null : (output as Record<string, unknown> | string | null),
            error,
          });

          // inbox.ask pauses until the human answers; feed the answer back
          // as the tool result so the loop continues on the reply.
          if (entry?.tool.name === "inbox.ask") {
            const answer = await runtime.waitForAnswer();
            if (answer == null) return; // session destroyed
            const text = answer.label ?? answer.body ?? answer.selectedChoiceId ?? "answered";
            messages.push({ role: "tool", tool_call_id: call.id, content: JSON.stringify({ answer: text }) });
          } else {
            messages.push({
              role: "tool",
              tool_call_id: call.id,
              content: JSON.stringify(error ? { error } : output),
            });
          }
        }
      }
      if (!aborted) fail("turn limit reached without finishing");
    };

    runLoop()
      .then(() => {
        /* outcome resolved inside runLoop */
      })
      .catch((err) => {
        if (!aborted) fail(err instanceof Error ? err.message : String(err));
      });

    return {
      runner: "deepseek",
      async injectReply(answer) {
        runtime.injectAnswer(answer);
      },
      async destroy() {
        aborted = true;
        runtime.cancelWait();
      },
      done,
    };
  }

  private async chat(
    baseUrl: string,
    model: string,
    messages: DeepseekMessage[],
    tools: unknown[],
  ): Promise<{
    choices?: { message?: { content?: string; tool_calls?: DeepseekToolCall[] } }[];
    usage?: { prompt_tokens?: number; completion_tokens?: number };
  }> {
    const res = await fetch(`${baseUrl}/chat/completions`, {
      method: "POST",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${this.opts.apiKey}`,
      },
      body: JSON.stringify({
        model,
        messages,
        tools: tools.length ? tools : undefined,
        tool_choice: tools.length ? "auto" : undefined,
        stream: false,
      }),
    });

    if (!res.ok) {
      const body = await res.text().catch(() => "");
      throw new Error(`DeepSeek API ${res.status}: ${body.slice(0, 500)}`);
    }
    return (await res.json()) as Awaited<ReturnType<DeepseekRunner["chat"]>>;
  }
}

File: agentos/runners/deepseek.py
"""
DeepSeek runner (BYOK, OpenAI-compatible API). Port of src/runners/deepseek.ts.
Uses httpx for async HTTP; tool calls loop mirrors the TS implementation.
"""
from __future__ import annotations

import asyncio
import json
from datetime import datetime, timezone
from typing import Any

import httpx

from agentos.mcp.index import servers_for_manifest
from agentos.runners.base import Runner, RunnerHandle, SessionOutcome
from agentos.runners.models import estimate_deepseek_cost


class _DeepSeekHandle(RunnerHandle):
    def __init__(self) -> None:
        self._done_fut: asyncio.Future[SessionOutcome] = asyncio.get_event_loop().create_future()
        self._runtime: Any = None

    @property
    def done(self):
        return self._done_fut

    async def inject_reply(self, answer: dict) -> None:
        if self._runtime:
            self._runtime.inject_answer(answer)

    async def destroy(self) -> None:
        if self._runtime:
            self._runtime.cancel_wait()


class DeepSeekRunner(Runner):
    kind = "deepseek"

    def __init__(self, api_key: str, base_url: str = "https://api.deepseek.com") -> None:
        self._api_key = api_key
        self._base_url = base_url.rstrip("/")

    async def provision(self, *, session_id: str, manifest: Any, runtime: Any,
                        cwd: str, script: Any = None) -> _DeepSeekHandle:
        handle = _DeepSeekHandle()
        handle._runtime = runtime
        asyncio.ensure_future(self._run(handle, session_id, manifest, runtime))
        return handle

    async def _run(self, handle: _DeepSeekHandle, session_id: str,
                    manifest: Any, runtime: Any) -> None:
        try:
            outcome = await self._invoke(session_id, manifest, runtime)
            if not handle._done_fut.done():
                handle._done_fut.set_result(outcome)
        except Exception as e:
            if not handle._done_fut.done():
                handle._done_fut.set_result(SessionOutcome(
                    status="failed", summary=None, cost_usd=None,
                    commit_shas=[], error=str(e),
                ))

    async def _invoke(self, session_id: str, manifest: Any, runtime: Any) -> SessionOutcome:
        servers = servers_for_manifest(manifest.mcpConnections)
        tools_schema = _build_tools_schema(servers)

        system = f"{manifest.agent.foundationalPrompt}\n\n{manifest.agent.rolePrompt}"
        if manifest.agent.skills:
            skill_text = "\n\n".join(s["body"] for s in manifest.agent.skills if s.get("body"))
            if skill_text:
                system += f"\n\n## Skills\n{skill_text}"

        user_msg = ""
        if manifest.task:
            user_msg = f"Task: {manifest.task.name}\n{manifest.task.description}"
        elif manifest.goal:
            user_msg = (
                f"Goal: {manifest.goal.title}\n{manifest.goal.spec}\n"
                f"Progress:\n{manifest.goal.progressLog or '(none)'}"
            )

        messages = [{"role": "user", "content": user_msg}]
        total_prompt_tokens = total_completion_tokens = 0

        async with httpx.AsyncClient(timeout=120) as client:
            for _ in range(50):  # max tool-call rounds
                resp = await client.post(
                    f"{self._base_url}/v1/chat/completions",
                    headers={"Authorization": f"Bearer {self._api_key}",
                              "Content-Type": "application/json"},
                    json={
                        "model": manifest.agent.model,
                        "messages": [{"role": "system", "content": system}, *messages],
                        "tools": tools_schema if tools_schema else None,
                        "tool_choice": "auto" if tools_schema else None,
                        "max_tokens": 4096,
                    },
                )
                resp.raise_for_status()
                data = resp.json()

                usage = data.get("usage", {})
                total_prompt_tokens += usage.get("prompt_tokens", 0)
                total_completion_tokens += usage.get("completion_tokens", 0)

                choice = data["choices"][0]
                msg = choice["message"]
                messages.append(msg)

                if choice["finish_reason"] == "tool_calls":
                    for tc in msg.get("tool_calls", []):
                        fn = tc["function"]
                        ts = datetime.now(timezone.utc).isoformat()
                        try:
                            args = json.loads(fn.get("arguments", "{}"))
                        except json.JSONDecodeError:
                            args = {}
                        tool_output = None
                        tool_error = None
                        try:
                            tool_output = await _dispatch_tool(servers, runtime, fn["name"], args)
                        except Exception as e:
                            tool_error = str(e)
                        await runtime.record_tool_call({
                            "ts": ts, "name": fn["name"], "input": args,
                            "output": tool_output, "error": tool_error,
                        })
                        messages.append({
                            "role": "tool",
                            "tool_call_id": tc["id"],
                            "content": json.dumps(tool_output) if tool_output is not None else (tool_error or ""),
                        })
                    continue

                # finish_reason == "stop" or other
                summary = msg.get("content") or ""
                cost = estimate_deepseek_cost(manifest.agent.model, total_prompt_tokens, total_completion_tokens)
                return SessionOutcome(status="ok", summary=summary[:500], cost_usd=cost, commit_shas=[])

        return SessionOutcome(status="failed", summary=None, cost_usd=None,
                               commit_shas=[], error="max tool-call rounds exceeded")


def _build_tools_schema(servers) -> list[dict]:
    tools = []
    for server in servers:
        for t in server.tools:
            tools.append({
                "type": "function",
                "function": {
                    "name": t.name.replace(".", "_"),
                    "description": t.description,
                    "parameters": {"type": "object", "properties": {}, "required": []},
                },
            })
    return tools


async def _dispatch_tool(servers, runtime, name: str, args: dict) -> Any:
    # DeepSeek uses underscores; MCP uses dots
    dot_name = name.replace("_", ".", 1)
    for server in servers:
        try:
            return await server.call(runtime, dot_name, args)
        except ValueError as e:
            if "unknown tool" in str(e):
                continue
            raise
    raise ValueError(f"unknown tool {name}")
